include/drivers/uart.h

```diff
@@ -38,6 +38,128 @@ enum uart_line_ctrl {
+ *
+ * Values that correspond to events or errors responsible for stopping
+ * receiving.
+ */
+enum uart_rx_stop_reason {
+	/** @brief Overrun error */
+	UART_ERROR_OVERRUN = (1 << 0),
+	/** @brief Parity error */
+	UART_ERROR_PARITY  = (1 << 1),
+	/** @brief Framing error */
+	UART_ERROR_FRAMING = (1 << 2),
+	/**
+	 * @brief Break interrupt
+	 *
+	 * A break interrupt was received. This happens when the serial input
+	 * is held at a logic '0' state for longer than the sum of
+	 * start time + data bits + parity + stop bits.
+	 */
+	UART_BREAK = (1 << 3),
+	/**
+	 * @brief Collision error
+	 *
+	 * This error is raised when transmitted data does not match
+	 * received data. Typically this is useful in scenarios where
+	 * the TX and RX lines maybe connected together such as
+	 * RS-485 half-duplex. This error is only valid on UARTs that
+	 * support collision checking.
+	 */
+	UART_ERROR_COLLISION = (1 << 4),
+};
+
+/** @brief Parity modes */
+enum uart_config_parity {
+	UART_CFG_PARITY_NONE,
+	UART_CFG_PARITY_ODD,
+	UART_CFG_PARITY_EVEN,
+	UART_CFG_PARITY_MARK,
+	UART_CFG_PARITY_SPACE,
+};
+
+/** @brief Number of stop bits. */
+enum uart_config_stop_bits {
+	UART_CFG_STOP_BITS_0_5,
+	UART_CFG_STOP_BITS_1,
+	UART_CFG_STOP_BITS_1_5,
+	UART_CFG_STOP_BITS_2,
+};
+
+/** @brief Number of data bits. */
+enum uart_config_data_bits {
+	UART_CFG_DATA_BITS_5,
+	UART_CFG_DATA_BITS_6,
+	UART_CFG_DATA_BITS_7,
+	UART_CFG_DATA_BITS_8,
+	UART_CFG_DATA_BITS_9,
+};
+
+/**
+ * @brief Hardware flow control options.
+ *
+ * With flow control set to none, any operations related to flow control
+ * signals can be managed by user with uart_line_ctrl functions.
+ * In other cases, flow control is managed by hardware/driver.
+ */
+enum uart_config_flow_control {
+	UART_CFG_FLOW_CTRL_NONE,
+	UART_CFG_FLOW_CTRL_RTS_CTS,
+	UART_CFG_FLOW_CTRL_DTR_DSR,
+};
+
+/**
+ * @brief UART controller configuration structure
+ *
+ * @param baudrate  Baudrate setting in bps
+ * @param parity    Parity bit, use @ref uart_config_parity
+ * @param stop_bits Stop bits, use @ref uart_config_stop_bits
+ * @param data_bits Data bits, use @ref uart_config_data_bits
+ * @param flow_ctrl Flow control setting, use @ref uart_config_flow_control
+ */
+struct uart_config {
+	uint32_t baudrate;
+	uint8_t parity;
+	uint8_t stop_bits;
+	uint8_t data_bits;
+	uint8_t flow_ctrl;
+};
+
+/**
+ * @defgroup uart_interrupt Interrupt-driven UART API
+ * @{
+ */
+
+/**
+ * @typedef uart_irq_callback_user_data_t
+ * @brief Define the application callback function signature for
+ * uart_irq_callback_user_data_set() function.
+ *
+ * @param dev       UART device structure.
+ * @param user_data Arbitrary user data.
+ */
+typedef void (*uart_irq_callback_user_data_t)(const struct device *dev,
+					      void *user_data);
+
+/**
+ * @typedef uart_irq_config_func_t
+ * @brief For configuring IRQ on each individual UART device.
+ *
+ * @param dev UART device structure.
+ *
+ * @internal
```

Comment:
```suggestion
```

Internal is not used to flag internal functions but an internal part of the docstring, ref. https://www.doxygen.nl/manual/commands.html#cmdinternal